.findExecutable <- function(exe, interactive=TRUE) {
  path <- Sys.which(exe)
  if(all(path=="")) {
    if(interactive) stop("Executable for ", paste(exe, collapse=" or "), " not found! Please make sure that the software is correctly installed and, if necessary, path variables are set.", call.=FALSE)
    return(character(0))
  }
  path[which(path!="")[1]]
}

datadownload  <- function(genome) {
  
  if (!file.exists(paste0("cytoBand.txt.gz"))) {
    cytoband <- paste0("curl -O http://hgdownload.cse.ucsc.edu/goldenPath/",genome,"/database/cytoBand.txt.gz")
    cat ("Running ",cytoband,"\n")
    system(cytoband, wait=T)
  }

  if (!file.exists(paste0(genome,".chrom.sizes"))) {
    genomesize <- paste0("curl -O http://hgdownload.cse.ucsc.edu/goldenPath/",genome,"/bigZips/",genome,".chrom.sizes")
    cat ("Running ",genomesize,"\n")
    system(genomesize, wait=T)
  }

  if (!file.exists(paste0(genome,".fa"))) {
    chromfa <- paste0("curl -O http://hgdownload.cse.ucsc.edu/goldenPath/",genome,"/bigZips/",genome,".fa.gz\n")
    cat ("Running ",chromfa,"\n")
    system(chromfa, wait=T)
  }

  if (!file.exists(paste0(genome,".refGene.gtf.gz"))) {
    gene <- paste0("curl -O http://hgdownload.cse.ucsc.edu/goldenPath/",genome,"/bigZips/genes/",genome,".refGene.gtf.gz")
    cat ("Running ",gene,"\n")
    system(gene, wait=T)
  }

} 

assingcompartment <- function(armbed, pcvec) {

  pcvec[pcvec <= 0] <- -1
  pcvec[pcvec  > 0] <- 1
  armbed[,"pcvec"]  <- pcvec

  if (nrow(armbed[armbed$arm=="p",]) > 0 ) { 
    p_arm_gc <- aggregate(gc ~ pcvec, mean, data=armbed[armbed$arm=="p",])
    p_arm_ts <- aggregate(trans ~ pcvec, mean, data=armbed[armbed$arm=="p",])
  
    if (nrow(p_arm_gc) == 1) {
      if (p_arm_gc$pcvec[1] == -1) {
        p_arm_gc[2,"pcvec"] <- 1
        p_arm_gc[2,"gc"] <- -1
      } else if (p_arm_gc$pcvec[1] == 1) {
        p_arm_gc[2,"pcvec"] <- -1
        p_arm_gc[2,"gc"] <- -1
      }
    }
    if (nrow(p_arm_ts) == 1) { 
      if (p_arm_ts$pcvec[1] == -1) { 
        p_arm_ts[2,"pcvec"] <- 1
        p_arm_ts[2,"trans"] <- -1
      } else if (p_arm_ts$pcvec[1] == 1) { 
        p_arm_ts[2,"pcvec"] <- -1
        p_arm_ts[2,"trans"] <- -1
      }
    }
  }
  
  if (nrow(armbed[armbed$arm=="q",]) > 0 ) {
    q_arm_gc <- aggregate(gc ~ pcvec, mean, data=armbed[armbed$arm=="q",])
    q_arm_ts <- aggregate(trans ~ pcvec, mean, data=armbed[armbed$arm=="q",])
    if (nrow(q_arm_gc) == 1) {
      if (q_arm_gc$pcvec[1] == -1) {
        q_arm_gc[2,"pcvec"] <- 1
        q_arm_gc[2,"gc"] <- -1
      } else if (q_arm_gc$pcvec[1] == 1) {
        q_arm_gc[2,"pcvec"] <- -1
        q_arm_gc[2,"gc"] <- -1
      }
    }
    if (nrow(q_arm_ts) == 1) {
      if (q_arm_ts$pcvec[1] == -1) {
        q_arm_ts[2,"pcvec"] <- 1
        q_arm_ts[2,"trans"] <- -1
      } else if (q_arm_ts$pcvec[1] == 1) {
        q_arm_ts[2,"pcvec"] <- -1
        q_arm_ts[2,"trans"] <- -1
      }
    }
  }

  if (nrow(armbed[armbed$arm=="p",]) > 0 & nrow(armbed[armbed$arm=="q",]) > 0) {
    d <- data.frame(arm=c("p","p","q","q"),sign=c(p_arm_gc$pcvec,q_arm_gc$pcvec),gc.mean=c(p_arm_gc$gc,q_arm_gc$gc),tss.mean=c(p_arm_ts$trans,q_arm_ts$trans))
    d[,"gc.response"] <- "Low"
    d[,"tss.response"] <- "Low"
    d[which.max(d[d$arm=="p",]$gc.mean),"gc.response"] <- "High"
    d[(2 + which.max(d[d$arm=="q",]$gc.mean)),"gc.response"] <- "High" 
    d[which.max(d[d$arm=="p",]$tss.mean),"tss.response"] <- "High"
    d[(2 + which.max(d[d$arm=="q",]$tss.mean)),"tss.response"] <- "High"
  } else if (nrow(armbed[armbed$arm=="p",]) == 0 & nrow(armbed[armbed$arm=="q",]) > 0) {
    d <- data.frame(arm=c("q","q"),sign=c(q_arm_gc$pcvec),gc.mean=c(q_arm_gc$gc),tss.mean=c(q_arm_ts$trans))
    d[,"gc.response"] <- "Low"
    d[,"tss.response"] <- "Low"
    d[which.max(d[d$arm=="q",]$gc.mean),"gc.response"] <- "High"
    d[which.max(d[d$arm=="q",]$tss.mean),"tss.response"] <- "High"
  } else if (nrow(armbed[armbed$arm=="p",]) > 0 & nrow(armbed[armbed$arm=="q",]) == 0) {
    d <- data.frame(arm=c("p","p"),sign=c(p_arm_gc$pcvec),gc.mean=c(p_arm_gc$gc),tss.mean=c(p_arm_ts$trans))
    d[,"gc.response"] <- "Low"
    d[,"tss.response"] <- "Low"
    d[which.max(d[d$arm=="p",]$gc.mean),"gc.response"] <- "High" 
    d[which.max(d[d$arm=="p",]$tss.mean),"tss.response"] <- "High"
  }
  d[d$gc.mean == -1,"gc.mean"] <- 0
  d[d$tss.mean== -1,"tss.mean"] <- 0
  return(d)
}


## obj : HMFA object file
## genome : hg38/hg19/mm9/mm10
## pc : component number, generally output from selectpc.r
## chr : chromosome name for which HMFA was ran
## resolution : HiC resolution
## choice : either tss (transcription start site) of gc (GC content)
## folder : path to the already existing _goldenpath folder generated by armcorrection. If you don't know, keep it as NA
armcorrection <- function(obj, genome, pc, chr, resolution, choice="tss", folder=NA) {
  
  if (is.na(folder)) {
    folder <- paste0(genome,"_goldenpathData")  
    if (!file.exists(folder)) {
      dir.create(folder)
    }
    folder <- normalizePath(folder)
  } else {
    folder <- normalizePath(folder)
  }
 
  current_path <- getwd()
  setwd(folder)
  datadownload(genome)
  setwd(current_path)

  if (file.exists(paste0(folder,"/",genome,".fa.gz"))) {
    cat ("Unzipping ",paste0(folder,"/",genome,".fa.gz"),"\n")
    system(paste0("gunzip ",genome,".fa.gz"), wait=T)
  }
  
  if (!file.exists(paste0(folder,"/",genome,".refGene.bed"))) {
    cmd <- paste0("gunzip -c ",folder,"/",genome,".refGene.gtf.gz |awk -v OFS='\\t' '{if($3==\"transcript\"){if($7==\"+\"){print $1,$4,$4+1}else{print $1,$5-1,$5}}}' |grep -v \"alt\" |grep -v \"random\" |sort -k 1,1 -k2,2n > ",folder,"/",genome,".refGene.bed")
    cat ("Running ",cmd,"\n")
    system(cmd, wait=T)
  } 
  
  if (!file.exists(paste0(folder,"/",genome,".binned.bed"))) {
    cmd <- paste0(.findExecutable("bedtools")," makewindows -g ",folder,"/",genome,".chrom.sizes -w ",as.integer(resolution)," > ",folder,"/",genome,".binned.bed")
    cat ("Running ",cmd,"\n")
    system(cmd, wait=T)
  }

  if (!file.exists(paste0(folder,"/",genome,".GCpt.bedGraph"))) {
    cmd <- paste0(.findExecutable("bedtools")," nuc -fi ",folder,"/",genome,".fa -bed ",folder,"/",genome,".binned.bed |grep -v \"#\" |awk -v OFS='\\t' '{print $1,$2,$3,$5}' |sort -k 1,1 -k2,2n > ",folder,"/",genome,".GCpt.bedGraph")
    cat ("Running ",cmd,"\n")
    system(cmd, wait=T)
  }

  cmd <- paste0("gunzip -c ",folder,"/cytoBand.txt.gz |grep -w \"",chr,"\" |awk -v OFS='\\t' '{if($4 ~ /p/){print $1,$2+1,$3,\"p\"}else{print $1,$2+1,$3,\"q\"}}' |",.findExecutable("bedtools")," groupby -i - -g 4 -c 1,2,3 -o distinct,min,max")
  cat ("Running ",cmd,"\n")
  arm_df <- read.table(text=system(cmd, wait=T, intern=T), h=F, as.is=T)
  colnames(arm_df) <- c("arm","chr","start","end")
  arm_df[,"start.index"] <- ceiling(arm_df$start/resolution)
  arm_df[,"end.index"]   <- ceiling(arm_df$end/resolution)
  print (arm_df)

  ## Change this to [.] bed <- do.call(rbind,strsplit(names(obj$quanti.var$coord[1:nrow(obj$ind[[1]]),1]),"[.]"))  
  bed <- do.call(rbind,strsplit(names(obj$quanti.var$coord[1:nrow(obj$ind[[1]]),1]),"[.]"))
  bed <- as.data.frame(bed, stringsAsFactors = F)
  colnames(bed) <- c("chr","start")
  bed$start <- as.integer(bed$start)
  bed[,"end"] <- bed$start + resolution
  bed$end <- as.integer(bed$end)
  bed[,"arm"] <- "q"
  bed[bed$start <= arm_df$end[1],"arm"] <- "p"
  bed <- bed[order(bed$start),]
  armfile <- tempfile(c(paste0(chr,".arms")), fileext=".bed")
  write.table(bed, file=armfile, row.names=F, col.names=F, sep="\t", quote=F)

  cmd <- paste0(.findExecutable("bedtools")," map -a ",armfile," -b ",folder,"/",genome,".GCpt.bedGraph -c 4 -o mean -null 0 |",.findExecutable("bedtools")," map -a - -b ",folder,"/",genome,".refGene.bed -c 1 -o count -null 0")
  cat ("Running ",cmd,"\n")
  armbed <- read.table(text=system(cmd, wait=T, intern=T), h=F)
  colnames(armbed) <- c("chr","start","end","arm","gc","trans")
  pcvector <- as.numeric(obj$ind$coord[,pc])
  compartment <- assingcompartment(armbed, pcvector)
  if (choice == "tss") {
    A.comp <- compartment[compartment$tss.response == "High",]
  } else if (choice == "gc") {
    A.comp <- compartment[compartment$gc.response == "High",]
  }

  n <- ncol(obj$partial[[1]][,pc,])
  for(k in 1:n) {
    pcvector <- as.numeric(obj$partial[[1]][,pc,k])
    armbed[,"pcvec"] <- pcvector
    if (nrow(armbed[armbed$arm == "p",]) > 0) {
      armbed[armbed$arm == "p",]$pcvec <- armbed[armbed$arm == "p",]$pcvec * A.comp[A.comp$arm == "p",]$sign
    }
    if (nrow(armbed[armbed$arm == "q",]) > 0) {
      armbed[armbed$arm == "q",]$pcvec <- armbed[armbed$arm == "q",]$pcvec * A.comp[A.comp$arm == "q",]$sign
    }
    obj$partial[[1]][,pc,k] <- armbed$pcvec
  }
  return(obj)
}

#obj <- readRDS("Replicate_Analysis/RDSdata/chr21.hmfaData.RDS")
#obj <- readRDS("/home/jeffreywang/overflow/DiffComptTesting/57CellLines/chr_22/Rsession.rds")
#armcorrection(obj=obj, genome="hg38", pc=2, chr="chr21", resolution=1e5, choice="tss", folder="hg38_goldenpathData")
